services/operations.go

```diff
@@ -0,0 +1,687 @@
+	ops []*types.Operation
+}
+
+// HasFee verifies whether the given operation list contains fee payment operations.
+func (m *operationToTransactionMapper) HasFee() bool {
+	return len(m.ops) >= 2 &&
+		m.ops[0].Type == OpTransfer &&
+		m.ops[1].Type == OpTransfer &&
+		m.ops[1].Account.Address == StringFromAddress(staking.FeeAccumulatorAddress)
+}
+
+// GetFee returns the fee (if any) extracted from the fee payment operations.
+//
+// If fee payment operations are present, this method also returns the signer address.
+func (m *operationToTransactionMapper) GetFee() (string, *transaction.Fee, error) {
+	fee := transaction.Fee{
+		Gas: DefaultGas,
+	}
+	if !m.HasFee() {
+		return "", &fee, nil
+	}
+
+	signWithAddr := m.ops[0].Account.Address
+	if m.ops[0].Account.SubAccount != nil {
+		return "", nil, fmt.Errorf("fee transfer from wrong subaccount (got: %s expected: nil)",
+			m.ops[0].Account.SubAccount,
+		)
+	}
+	feeAmount, err := readOasisCurrencyNeg(m.ops[0].Amount)
+	if err != nil {
+		return "", nil, fmt.Errorf("invalid fee transfer from amount: %w", err)
+	}
+	if feeGasRaw, ok := m.ops[0].Metadata[FeeGasKey]; ok {
+		feeGasF64, ok := feeGasRaw.(float64)
+		if !ok {
+			return "", nil, fmt.Errorf("malformed fee transfer gas metadata")
+		}
+		fee.Gas = transaction.Gas(feeGasF64)
+	}
+
+	if m.ops[1].Account.SubAccount != nil {
+		return "", nil, fmt.Errorf("fee transfer to wrong subaccount (got: %s expected: nil)",
+			m.ops[1].Account.SubAccount,
+		)
+	}
+	feeAmount2, err := readOasisCurrency(m.ops[1].Amount)
+	if err != nil {
+		return "", nil, fmt.Errorf("invalid fee transfer to amount: %w", err)
+	}
+	if feeAmount.Cmp(feeAmount2) != 0 {
+		return "", nil, fmt.Errorf("fee transfer amounts differ between operations (from: %s to: %s)", feeAmount, feeAmount2)
+	}
+	fee.Amount = *feeAmount
+	return signWithAddr, &fee, nil
+}
+
+func readCurrency(amount *types.Amount, currency *types.Currency, negative bool) (*quantity.Quantity, error) {
+	// TODO: Is it up to us to check other fields?
+	if amount.Currency.Symbol != currency.Symbol {
+		return nil, fmt.Errorf("wrong currency")
+	}
+	bi := new(big.Int)
+	if err := bi.UnmarshalText([]byte(amount.Value)); err != nil {
+		return nil, fmt.Errorf("bi UnmarshalText Value: %w", err)
+	}
+	if negative {
+		bi.Neg(bi)
+	}
+	q := quantity.NewQuantity()
+	if err := q.FromBigInt(bi); err != nil {
+		return nil, fmt.Errorf("q FromBigInt bi: %w", err)
+	}
+	return q, nil
+}
+
+func readOasisCurrency(amount *types.Amount) (*quantity.Quantity, error) {
+	return readCurrency(amount, OasisCurrency, false)
+}
+
+func readOasisCurrencyNeg(amount *types.Amount) (*quantity.Quantity, error) {
+	return readCurrency(amount, OasisCurrency, true)
+}
+
+// GetTransaction decodes an Oasis transaction from given Rosetta operations.
+//
+// The method also returns the signer address.
+func (m *operationToTransactionMapper) GetTransaction() (string, *transaction.Transaction, error) {
+	signWithAddr, fee, err := m.GetFee()
```

Comment:
:information_source: in the case that there is no fee, the logic for determining signWithAddr is enhanced, now with routines specific to each transaction type